File: src/main/java/org/mitallast/queue/raft/protocol/InstallSnapshot.java
package org.mitallast.queue.raft.protocol;

import org.mitallast.queue.common.codec.Codec;
import org.mitallast.queue.common.codec.Message;
import org.mitallast.queue.transport.DiscoveryNode;

public class InstallSnapshot implements Message {
    public static final Codec<InstallSnapshot> codec = Codec.Companion.of(
        InstallSnapshot::new,
        InstallSnapshot::getLeader,
        InstallSnapshot::getTerm,
        InstallSnapshot::getSnapshot,
        DiscoveryNode.codec,
        Codec.Companion.longCodec(),
        RaftSnapshot.codec
    );

    private final DiscoveryNode leader;
    private final long term;
    private final RaftSnapshot snapshot;

    public InstallSnapshot(DiscoveryNode leader, long term, RaftSnapshot snapshot) {
        this.leader = leader;
        this.term = term;
        this.snapshot = snapshot;
    }

    public DiscoveryNode getLeader() {
        return leader;
    }

    public long getTerm() {
        return term;
    }

    public RaftSnapshot getSnapshot() {
        return snapshot;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        InstallSnapshot that = (InstallSnapshot) o;

        if (term != that.term) return false;
        if (!leader.equals(that.leader)) return false;
        return snapshot.equals(that.snapshot);
    }

    @Override
    public int hashCode() {
        int result = leader.hashCode();
        result = 31 * result + (int) (term ^ (term >>> 32));
        result = 31 * result + snapshot.hashCode();
        return result;
    }

    @Override
    public String toString() {
        return "InstallSnapshot{" +
            "leader=" + leader +
            ", term=" + term +
            ", snapshot=" + snapshot +
            '}';
    }
}

File: src/main/java/org/mitallast/queue/raft/protocol/InstallSnapshot.kt
package org.mitallast.queue.raft.protocol

import org.mitallast.queue.common.codec.Codec
import org.mitallast.queue.common.codec.Message
import org.mitallast.queue.transport.DiscoveryNode

data class InstallSnapshot(val leader: DiscoveryNode, val term: Long, val snapshot: RaftSnapshot) : Message {
    companion object {
        val codec = Codec.of(
            ::InstallSnapshot,
            InstallSnapshot::leader,
            InstallSnapshot::term,
            InstallSnapshot::snapshot,
            DiscoveryNode.codec,
            Codec.longCodec(),
            RaftSnapshot.codec
        )
    }
}
